localstack/services/secretsmanager/provider.py

```diff
@@ -0,0 +1,357 @@
+    ClientRequestTokenType,
+    CreateSecretResponse,
+    DeleteResourcePolicyResponse,
+    DeleteSecretResponse,
+    DescribeSecretResponse,
+    DescriptionType,
+    GetResourcePolicyResponse,
+    GetSecretValueResponse,
+    KmsKeyIdType,
+    ListSecretVersionIdsResponse,
+    MaxResultsType,
+    NameType,
+    NextTokenType,
+    NonEmptyResourcePolicyType,
+    PutResourcePolicyResponse,
+    PutSecretValueResponse,
+    RecoveryWindowInDaysType,
+    RemoveRegionsFromReplicationResponse,
+    RemoveReplicaRegionListType,
+    ReplicateSecretToRegionsResponse,
+    RestoreSecretResponse,
+    RotateSecretResponse,
+    RotationLambdaARNType,
+    RotationRulesType,
+    SecretBinaryType,
+    SecretIdType,
+    SecretsmanagerApi,
+    SecretStringType,
+    SecretVersionIdType,
+    SecretVersionStagesType,
+    SecretVersionStageType,
+    StopReplicationToReplicaResponse,
+    TagKeyListType,
+    TagListType,
+    UpdateSecretResponse,
+    UpdateSecretVersionStageResponse,
+    ValidateResourcePolicyResponse,
+)
+from localstack.constants import TEST_AWS_ACCOUNT_ID
+from localstack.services.moto import call_moto, call_moto_with_request
+from localstack.utils.aws import aws_stack
+from localstack.utils.common import to_str
+
+LOG = logging.getLogger(__name__)
+
+
+class SecretsmanagerProvider(SecretsmanagerApi, ABC):
+    def __init__(self):
+        super().__init__()
+        apply_patches()
+
+    @staticmethod
+    def _transform_context_secret_id(context: RequestContext) -> Optional[dict]:
+        data_dict = json.loads(to_str(context.request.data or "{}"))
+        secret_id = data_dict.get("SecretId", None)
+        if secret_id and ":" in secret_id:
+            parts = secret_id.split(":")
+            if parts[3] != aws_stack.get_region():
+                LOG.info(
+                    'Unexpected request region %s for secret "%s"',
+                    aws_stack.get_region(),
+                    secret_id,
+                )
+            # secret ARN ends with "-<randomId>" which we remove in the request for upstream compatibility
+            # if the full arn is being sent then we remove the string in the end
+            if parts[-1][-7] == "-":
+                secret_id = parts[-1][: len(parts[-1]) - 7]
+            elif parts[-1][-1] != "-":
+                secret_id = secret_id + "-"
+            #
+            data_dict["SecretId"] = secret_id
+            return data_dict
+        return None
+
+    @staticmethod
+    def _call_moto_with_request_secret_id(context: RequestContext) -> ServiceResponse:
+        data_dict = SecretsmanagerProvider._transform_context_secret_id(context)
+        return call_moto_with_request(context, data_dict) if data_dict else call_moto(context)
+
+    def cancel_rotate_secret(
+        self, context: RequestContext, secret_id: SecretIdType
+    ) -> CancelRotateSecretResponse:
+        return CreateSecretResponse(**self._call_moto_with_request_secret_id(context))
```

Comment:
i think all of these can be simplified by `return self._call_moto_with_request_secret_id(context))`, since 
`d == dict(**d)`